File: API/Services/CourierService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using API.Couriers;
using API.DTOs;
using API.Entities;
using API.Interfaces;

namespace API.Services
{
    public class CourierService : ICourierService
    {
        List<Courier> Couriers;
        private readonly IOrderRepository _orderRepository;

        public CourierService(IOrderRepository orderRepository)
        {
            _orderRepository = orderRepository;
            Couriers = new List<Courier>
            {
                new Cargo4You("Cargo4You"),
                new ShipFaster("ShipFaster"),
                new MaltaShip("MaltaShip")
            };
        }


        public ServiceResponse<List<Courier>> GetCourierPrices(PackageDto packageDto)
        {
            List<Courier> couriers = new();

            foreach (Courier c in Couriers)
            {
                if (c.ValidateDimension(packageDto) && c.ValidateWeight(packageDto))
                {
                    c.CalculatePrice(packageDto);
                    couriers.Add(c);
                }
            }
            return new ServiceResponse<List<Courier>>{ Data = couriers };
        }

        public async Task<ServiceResponse<bool>> MakeOrder(OrderDto orderDto)
        {
             var order = new Order
            {
                Package = new Package
                {
                    Weight = orderDto.PackageDto.Weight,
                    Height = orderDto.PackageDto.Height,
                    Width = orderDto.PackageDto.Width,
                    Depth = orderDto.PackageDto.Depth
                },
                OrderDate = DateTime.Now,
                CourierName = orderDto.CourierName,
                CourierPrice = orderDto.CourierPrice,
            };

            await _orderRepository.MakeOrder(order);
            return new ServiceResponse<bool>{ Data = true, Message = "Order placed successfully" };
        }
    }
}

File: client/src/app/_services/courier.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Parcel } from '../_models/parcel';
import { Courier } from '../_models/courier';
import { Order } from '../_models/order';
import { ServiceResponse } from '../_models/serviceResponse';

@Injectable({
  providedIn: 'root'
})
export class CourierService {
  baseUrl = 'https://localhost:5001/api/';

  constructor(private http: HttpClient) { }

  getCourierPrices(parcel: Parcel){
    let params = new HttpParams();
    params = params.append("weight", parcel.weight)
    params = params.append("height", parcel.height)
    params = params.append("width", parcel.width)
    params = params.append("depth", parcel.depth)
    
    return this.http.get<ServiceResponse<Courier[]>>(this.baseUrl + 'shipping', { params: params });
  }

  makeOrder(order: Order){
    return this.http.post<ServiceResponse<boolean>>(this.baseUrl + 'shipping', order);
  }
}
